File: 376-wiggle-subsequence/376-wiggle-subsequence.cpp
class Solution {
public:
    int wiggleMaxLength(vector<int>& nums) {
        // Variation of longest increasing subsequence (albeit abit slow)
        // Longest Increasing Subsequence: https://leetcode.com/problems/longest-increasing-subsequence/
        int n = nums.size();
        vector<int> wiggleDown(n, 1);
        vector<int> wiggleUp(n, 1);
        int longest = 1;
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[j] < nums[i]) wiggleUp[i] = max(wiggleUp[i], wiggleDown[j]+1);
                else if (nums[j] > nums[i]) wiggleDown[i] = max(wiggleDown[i], wiggleUp[j]+1);
            }
            int maxWiggle = max(wiggleDown[i], wiggleUp[i]);
            longest = max(longest, maxWiggle);
        }
        return longest;
    }
};

File: 376-wiggle-subsequence/376-wiggle-subsequence.py
class Solution(object):
    def wiggleMaxLength(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # Extended Variation of Longest Increasing Subsequence
        # https://leetcode.com/problems/longest-increasing-subsequence/
        longest_size = 1
        dp_positive = [1 for _ in range(len(nums))]
        dp_negative = [1 for _ in range(len(nums))]
        for i in range(1, len(nums)):
            for j in range(i):
                if (nums[j] < nums[i]): # For increment
                    # Compare current max against previous max + 1
                    # Previous max in this case must be from the decrement array -- dp_negative
                    dp_positive[i] = max(dp_positive[i], dp_negative[j]+1)
                    
                if (nums[j] > nums[i]): # For decrement
                    # Compare current max against previous max + 1
                    # Previous max in this case must be from the increment array -- dp_positive
                    dp_negative[i] = max(dp_negative[i], dp_positive[j]+1)
            longest_size = max(longest_size, max(dp_positive[i], dp_negative[i]))
        return longest_size